src/paillier.rs

```diff
@@ -86,11 +105,21 @@ impl PaillierEncryptionKey {
 pub(crate) struct PaillierDecryptionKey(libpaillier::DecryptionKey);
 
 impl PaillierDecryptionKey {
-    pub(crate) fn decrypt(&self, c: &PaillierCiphertext) -> Result<Vec<u8>> {
-        Ok(self
+    pub(crate) fn decrypt(&self, c: &PaillierCiphertext) -> Result<BigNumber> {
+        let mut x = self
             .0
             .decrypt(&c.0)
-            .ok_or(PaillierError::DecryptionFailed)?)
+            .ok_or(PaillierError::DecryptionFailed)
+            .map(BigNumber::from_slice)?;
+
+        // Switch representation into `[-N/2, N/2]`. libpaillier (and indeed, `BigNumber`s
+        // in general) returns values represented in the canonical range [0, N). A single
+        // subtraction will land us in the expected range for this application.
+        if x > (self.0.n() - 1) / 2 {
```

Comment:
If the above `half_n` suggestion is taken, we can replace this with `x > self.half_n`

Reply:
Sadly this is on `DecryptionKey`, not `EncryptionKey`, so I made another function to do the same thing for this type.